src/EditorFeatures/Test/Snippets/RoslynLSPSnippetConvertTests.cs

```diff
@@ -0,0 +1,243 @@
+@"$0 if (true)
+{
+} ${1:test}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForPlaceholderForwardsForCaret()
+        {
+            var markup =
+@"{|placeholder:test|} [|if (true)
+{
+}|] $$";
+
+            var expectedLSPSnippet =
+@"${1:test} if (true)
+{
+} $0";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        #endregion
+
+        #region LSP Snippet generation tests
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestForLoopSnippet()
+        {
+            var markup =
+@"[|for (var {|placeholder1:i|} = 0; {|placeholder1:i|} < {|placeholder2:length|}; {|placeholder1:i|}++)
+{$$
+}|]";
+
+            var expectedLSPSnippet =
+@"for (var ${1:i} = 0; ${1:i} < ${2:length}; ${1:i}++)
+{$0
+}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        #endregion
+
+        protected static TestWorkspace CreateWorkspaceFromCode(string code)
+         => TestWorkspace.CreateCSharp(code);
+
+        private static async Task TestAsync(string markup, string output)
+        {
+            MarkupTestFile.GetPositionAndSpans(markup, out var text, out var cursorPosition, out IDictionary<string, ImmutableArray<TextSpan>> placeholderDictionary);
+            var stringSpan = placeholderDictionary[""].First();
+            var textChange = new TextChange(new TextSpan(stringSpan.Start, 0), text.Substring(stringSpan.Start, stringSpan.Length));
+            var placeholders = GetSnippetPlaceholders(text, placeholderDictionary);
+            using var workspace = CreateWorkspaceFromCode(markup);
+            var document = workspace.CurrentSolution.GetDocument(workspace.Documents.First().Id);
+
+            var lspSnippetString = await RoslynLSPSnippetConverter.GenerateLSPSnippetAsync(document!, cursorPosition!.Value, placeholders, textChange, CancellationToken.None).ConfigureAwait(false);
+            AssertEx.EqualOrDiff(output, lspSnippetString);
+        }
+
+        private static ImmutableArray<SnippetPlaceholder> GetSnippetPlaceholders(string text, IDictionary<string, ImmutableArray<TextSpan>> placeholderDictionary)
+        {
+            using var _ = ArrayBuilder<SnippetPlaceholder>.GetInstance(out var arrayBuilder);
+            foreach (var kvp in placeholderDictionary)
+            {
+                if (kvp.Key.Length > 0)
+                {
+                    var spans = kvp.Value;
+                    var identifier = text.Substring(spans[0].Start, spans[0].Length);
+                    var placeholders = spans.Select(span => span.Start).ToImmutableArray();
+                    arrayBuilder.Add(new SnippetPlaceholder(identifier, placeholders));
+                }
+            }
+
+            return arrayBuilder.AsImmutable();
```

Comment:
```suggestion
            return arrayBuilder.ToImmutable();
```

we should have an analyzer catch this :D